packages/react-dom/src/server/ReactPartialRendererHooks.js

```diff
@@ -481,12 +484,91 @@ function useTransition(): [(callback: () => void) => void, boolean] {
   return [startTransition, false];
 }
 
+function makeNestedOpaqueHydratingObjectFromKey(
+  parent: OpaqueIDType,
+  getValueOfKey: () => string,
+): OpaqueIDType {
+  const keys: {
+    unsorted: Array<string>,
+    map: Map<string, number>,
+  } = {
+    unsorted: [],
+    map: new Map(),
+  };
+  return {
+    makeNewFromKey(key: string) {
+      keys.unsorted.push(key);
+      return makeNestedOpaqueHydratingObjectFromKey(this, () => {
+        if (keys.map.size !== keys.unsorted.length) {
+          keys.map = new Map();
+          keys.unsorted.sort((a: string, b: string): number => {
+            return a.length === b.length
+              ? a.localeCompare(b)
+              : a.length - b.length;
+          });
+          keys.unsorted.forEach((element, index) => {
+            keys.map.set(element, index);
+          });
+        }
+        const index = keys.map.get(key);
+        invariant(
+          typeof index === 'number',
+          'The key was not found in the index map in makeNewFromKey. This is a bug in react.',
+        );
+        return index.toString(36);
+      });
+    },
+    toString() {
+      return parent.toString() + ':' + getValueOfKey();
+    },
+    valueOf() {
+      return parent.valueOf() + ':' + getValueOfKey();
+    },
+  };
+}
+
 function useOpaqueIdentifier(): OpaqueIDType {
-  return (
+  const id =
     (currentPartialRenderer.identifierPrefix || '') +
     'R:' +
-    (currentPartialRenderer.uniqueID++).toString(36)
-  );
+    (currentPartialRenderer.uniqueID++).toString(36);
+  const keys: {
+    unsorted: Array<string>,
+    map: Map<string, number>,
+  } = {
+    unsorted: [],
+    map: new Map(),
+  };
+  return {
+    makeNewFromKey(key: string) {
```

Comment:
Is this section (543<563) the same as 499->527?  If so, move to a shared function?